### ISSUE TITLE
Less memory copies with Memory.MapSlice(offset, []byte)

### ISSUE THREAD

**Issue by codefromthecrypt** [tetratelabs/wazero#525]
Folks are doing more work moving data into the memory sandbox addressable by Wasm.

For example, @knqyf263 is working on [structured encoding][1] and @inkeliz is working on [random-access encoding][2]. Structs aside, the problem exists even with strings. You need to pass them into wasm (currently a copy) and its sandbox shouldn't be violated in the process..

Some alternatives solve this by providing unsafe access to the entire memory (ex a []byte). However, doing this breaks the sandbox and also any access controls you may have around boundaries. Furthermore, it may confuse who owns allocation, and typically that's the wasm code itself (ex TinyGo or rust malloc/free functions). Finally, it cements the underlying structure of memory, making changing it, ex [to pages of byte arrays][3], difficult later.

Another way could be to scope a memory overlay, given an offset and a range. For example, you could call an exported function to allocate memory, then apply an overlay which reads from that instead of the underlying source.

```go
var thing []byte = // something to manage in go

// Reserve linear memory using ex TinyGo's allocator.
results, _ := malloc.Call(ctx, len(thing))
ptr := results[0]
defer free.Call(ctx, ptr)

// map the slice so that until unmap is called, any writes to this region are write-through regardless
// of whether it is updated by Go or Wasm. Since this block is after malloc and before free, TinyGo
// shouldn't attempt to clear it.
unmap := mod.Memory().MapSlice(ptr, thing)
defer unmap()
```

Note this pattern could be accomplished many ways. For example if there is global mutable state,
the size could be indicated by a Global which is allocated during an "init" function. The actual
mapping could be done in a host function, which makes it possible to do in a .start function.
Regardless of how this is done, it could allow more precise handing over control of a region to
the host vs exposing the entire memory.

To do this internally may be trickier as there could be many stacks of these layers, so I'm not
entirely sure we want to even do this (think JIT). However, I figured we can think about it a while
and gather feedback about options.

[1]: #523
[2]: https://github.com/tetratelabs/wazero/issues/523#issuecomment-1117634602
[3]: #288

**Comment by codefromthecrypt** [comment]
One way to make this relatively easy would be to force such write-through regions to be fixed size (a page or 64KB). Provided you can control the often compiled-in memory limits, this would be neat. Ex. Memory.UnsafePage(offset)

**Comment by codefromthecrypt** [comment]
though even if we did that it wouldn't necessarily end up page aligned.. in any case it would be a coarser tool and encourage folks to not do this for chatty small amounts. anyway food for thought.

**Comment by codefromthecrypt** [comment]
Also, I'm wrong about stacks of these layers as that would be impossible unless the allocator had a bug. each mapped slice or reserved page could only have overlaps if allocation was broken.

**Comment by inkeliz** [comment]
I'm not sure if I understand. I think `mod.Memory().Read(nil, offset, size)` already does this job, and doesn't copy the content. Well, as long as the buffer doesn't allocates (because in that case, the pointer is lost).

Again, I'm not 100% if I understand the issue. 

-------------

Consider that WASM:

```
var buffer []byte

//export Output
func Output(i uint32) uint32 {
	buffer = make([]byte, i)
	return uint32(uintptr(unsafe.Pointer(&buffer[0])))
}

//export Update
func Update() {
	*(*uint32)(unsafe.Pointer(&buffer[0]))++
}
```

Then, you can use Wazero to call both functions:

```
func TestCopy(t *testing.T) {
	runtime := wazero.NewRuntimeWithConfig(wazero.NewRuntimeConfigJIT().WithFinishedFeatures())
	wasi.InstantiateSnapshotPreview1(context.Background(), runtime)

	wasifile, _ := os.ReadFile(filepath.Join("testdata", "wasi", "inc.wasi"))
	wasm, _ := runtime.CompileModule(context.Background(), wasifile)
	m, _ := runtime.InstantiateModuleWithConfig(context.Background(), wasm, wazero.NewModuleConfig().WithStdout(os.Stdout).WithStderr(os.Stdout))

	type Counter struct {
		Points uint32
	}

	// call exported function
	outputPtr, err := m.ExportedFunction("Output").Call(context.Background(), uint64(unsafe.Sizeof(Counter{})))
	if err != nil {
		t.Fatal(err)
	}

	// reads memory (it doesn't copy)
	output, ok := m.Memory().Read(nil, uint32(outputPtr[0]), uint32(unsafe.Sizeof(Counter{})))
	if !ok {
		t.Fatal("impossible memory")
	}

	// force it to be one "Counter"
	counter := (*Counter)(unsafe.Pointer(&output[0]))

	for i := uint32(1); i < 1000; i++ {
		// call exported function, which increeases the counter
		m.ExportedFunction("Update").Call(context.Background())

		// reads from the mapped stuff
		if i != counter.Points {
			t.Fatal("error")
		}
		fmt.Println(counter.Points)
	}
}
```

The test will PASS (--- PASS: TestCopy). Notice, that I'm only setting `counter := (*Counter)(unsafe.Pointer(&output[0]))` once, and then reading it inside of the `for`. 
 
 This code will break if TinyGo grow their memory and cause `append` to alloc. However, you can use do:
 ```
 	output, ok := m.Memory().Read(nil, uint32(outputPtr[0]), uint32(unsafe.Sizeof(Counter{})))
	if !ok {
		t.Fatal("impossible memory")
	}
	counter := (*Counter)(unsafe.Pointer(&output[0]))
```
Inside of the loop. That is not copying anything, and it's basically one viewer/overlay of the bytes, as you described before. So, I'm not sure if I understand the idea, or I'm missing something.

**Comment by codefromthecrypt** [comment]
I see what you mean and thanks for thinking about it and giving an example.

The main point was thinking about an struct view backed by a slice, Ex. `Animal` where `Name()` reads from a slice offset. At the time I was thinking to instantiate `Animal` and give wazero its slice (ex `Animal.Data()`). On the other hand, what you said works too, you can do the same thing like `NewAnimal(rawSlice)` or use `reflect.SliceHeader` to assign it from a slice wazero provides (ex via `Memory.Read`).

The example, I was thinking of was getting a bit away from go's internal representation of structs and more on a random-access serialization type, but your example is also valid... the main point is how we get the effect of write-through without any copying.

I had forgotten that we implement Memory.Read as [re-slicing](https://github.com/tetratelabs/wazero/blob/main/internal/wasm/memory.go#L125), so agree with your points above that this should work. That would work as long as TinyGo doesn't try to collect the region, which could be achieved by externally using the exported allocation function (as my example did), or the value passed back is kept alive in TinyGo (ex by assigning it to a `var` or a map value etc.

One trick is that we've no tests for the Memory.Read side effect, even if it happens to work now. Ex to keep this, we should add tests and also check behavior after `Grow` (luckily there is no shrink). We may end up with an issue if we switch the impl underneath to use managed pages instead of go's slice impl, but otoh there's no current requirement to do that either.

Make sense?

So, TL;DR; is that your approach above is clever, and I would guess even without executing your tests that it works. We should backfill tests to know it works post memory.grown, and it keeps working. It is a little hacky to rely on how Memory.Read is implemented without this.

**Comment by codefromthecrypt** [comment]
the other main point we should note is that folks thinking like this should forget about `Memory.Write` if they can redesign their sharing approach to exclusively us `Memory.Read`, it can be viable. Check me if that's a bad way of putting it.

**Comment by codefromthecrypt** [comment]
another note for how we test this is that we can use benchmarks that track allocation. While CGO based solutions are tricky to figure out memory usage, this is one area we can leverage greatly. If an approach says it doesn't result in allocation (above a baseline) it is fairly straightforward to validate.

**Comment by codefromthecrypt** [comment]
What I'll do next is raise a PR on this and ask for your comments @inkeliz. It might be better off as an adhoc test vs an example as there are a couple patterns to verify.

**Comment by codefromthecrypt** [comment]
ok I finished up https://github.com/tetratelabs/wazero/pull/527 which should close this out. I think documentation and tests on Memory.Read is better than adding a new API. Thanks for the idea @inkeliz
